### ISSUE TITLE
Store MSAL logged-in users and precheck them before doing a request on node.js web app

### ISSUE THREAD

**Issue by kwoxer** [AzureAD/microsoft-authentication-library-for-js#3067]
## Library
- [ ] `msal@1.x.x` or `@azure/msal@1.x.x`
- [ ] `@azure/msal-browser@2.x.x`
- [x] `@azure/msal-node@1.0.0`
- [ ] `@azure/msal-react@1.x.x`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angular@2.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

## Description

In my node project I have the following basic code to connect to Azure via a token. The login/logout works great together with our Azure:

    const express = require("express");
    const msal = require('@azure/msal-node');
    const SERVER_PORT = process.env.PORT || 3000;
    const config = {
        auth: {
            clientId: "XXX",
            authority: "https://login.microsoftonline.com/common",
            clientSecret: "XXX"
        },
        system: {
            loggerOptions: {
                loggerCallback(loglevel, message, containsPii) {
                    console.log(message);
                },
                piiLoggingEnabled: false,
                logLevel: msal.LogLevel.Verbose,
            }
        }
    };
    const pca = new msal.ConfidentialClientApplication(config);
    const app = express();
    
    app.get('/', (req, res) => {
        res.send("<a href=\"login\">Login</a> <a href=\"logout\">Logout</a>");
    });
    app.get('/dashboard', (req, res) => {
        // check here for valid token...
    });
    app.get('/login', (req, res) => {
        const authCodeUrlParameters = {
            scopes: ["user.read"],
            redirectUri: "http://localhost:3000/redirect",
        };
        pca.getAuthCodeUrl(authCodeUrlParameters).then((response) => {
            res.redirect(response);
        }).catch((error) => console.log(JSON.stringify(error)));
    });
    app.get('/logout', (req, res) => {
        res.redirect('https://login.microsoftonline.com/common/oauth2/v2.0/logout?post_logout_redirect_uri=http://localhost:3000/');
    });
    app.get('/redirect', (req, res) => {
        const tokenRequest = {
            code: req.query.code,
            scopes: ["user.read"],
            redirectUri: "http://localhost:3000/redirect",
        };
        pca.acquireTokenByCode(tokenRequest).then((response) => {
            console.log("\nResponse: \n:", response);
            res.sendStatus(200);
        }).catch((error) => {
            console.log(error);
            res.status(500).send(error);
        }); 
    });
    app.listen(SERVER_PORT, () => console.log(`Msal Node Auth Code Sample app listening on port ${SERVER_PORT}!`))

But how to properly check after that logging in if the token is still valid?

So the question is, how can I be save that the user on /dashboard has still a valid token or is logged in?

    app.get('/dashboard', (req, res) => {
        // check here for valid token...
    });

At the end I need a node.js application that:
- is safe (token-based and secrets stay on the node.js server)
- user only logs in **once** until he logout on its own
- has user auth (msal)
- can give granular permissions on routes

Can I do all that in node.js or better doing that in client-side? But am I then reducing the security? Or do I even need an extention like this https://www.npmjs.com/package/@azure/msal-node-extensions and https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/extensions/samples/msal-node-extensions/index.js?

## Source

- [ ] Internal (Microsoft)
- [x] Customer request

**Comment by pkanher617** [comment]
@kwoxer Yes you should be able to do the things that you are describing above.

In our [E2E sample](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/EndToEndTestApp), we demonstrate how to make a graph call using an account API or `acquireTokenSilent`, which will ensure that you have a logged in user, and that the token you retrieve is valid (not expired). 

```
// Initiates Acquire Token Silent flow
  app.get('/graphCall', async (req, res) => {
      // get Accounts
      const account = await msalTokenCache.getAccountByHomeId(app.locals.homeAccountId);
      /** 
       * Account index must match the account's position in the cache. The sample cache file contains a dummy account
       * entry in index 0, hence the actual account that is logged in will be index 1
       */
      const silentRequest = { ...requestConfig.silentRequest, account: account };
  
      let templateParams = { showLoginButton: false };
      // Acquire Token Silently to be used in MS Graph call
      clientApplication.acquireTokenSilent(silentRequest)
          .then((authResponse) => {
              handleAuthorizationResponse(res, authResponse, templateParams, scenarioConfig);
          })
          .catch((error) => {
              console.log(error);
              templateParams.couldNotAcquireToken = true;
              res.render("graph", templateParams)
          });
  });
```

> - is safe (token-based and secrets stay on the node.js server)
> - user only logs in once until he logout on its own
> - has user auth (msal)
> - can give granular permissions on routes

- all MSAL account information and tokens stay in memory by default, you can control this by implementing the cache APIs. There is a [file](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/msal-node-samples/EndToEndTestApp/app/cachePlugin.js) in the sample linked above that should give you an idea.
- This can be controlled by making sure you keep track of a single user, and do not allow another user to sign in until the single user has signed out of you application. The silent flow should also make sure you do not have to sign-in again until the user logs out of the application and the cache is cleared.
- You are using the auth code flow, which should be authenticating your user and giving you an access token and an id token for the user.
- The permissions per route will need to be handled by your app, but you can use the `acquireTokenSilent()` method to fetch a valid token from the cache and ensure the user has access to the permissions you need for that route.

> Or do I even need an extention like this https://www.npmjs.com/package/@azure/msal-node-extensions and https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/extensions/samples/msal-node-extensions/index.js?

The extensions library is meant to support desktop application to save their cache to disk across platforms. It isn't recommended for web applications.

**Comment by kwoxer** [comment]
Thanks for the ideas and thoughts. It's so confusing for me to have so many different examples and now to pick the best that fits in my case. But I still do not understand at all when to choose the **node** version and when to choose the **browser** version.

My idea was to create the token on node, because then I can hide the `clientSecret` from the user. But when I see the examples there sometimes the `clientSecret` is also delievered to the browser to be run on client-side. Isn't this a security issue? Or is the `clientSecret` not secret at all?

So as I understand you, node is not needed and I should write my whole code in browser? Or did I get you wrong?

As I understood the security chain, I wanted to create the token with the code from above. Then the user is redirected to `/redirect`. There I set the recieved token to `localStorage` and then I check on any further request that token for still being valid. But what I do not understand here is that the whole security check is again on client-side. So anyone is able to simply manipulate the javascript to get access. Or maybe I did not get the point of all. I have no idea where to start and what to test.

So the basics on node work but I need help on the whole concept if you have time to explain it to me. 

And I tried to run the test project but it failed:
 
![image](https://user-images.githubusercontent.com/1167229/108972452-9eaf1600-7683-11eb-91b3-0c92c06765ca.png)

**Comment by kwoxer** [comment]
Ahh cool I found a kind of dirty way to solve my issue. Could you maybe tell me if that is a proper way? Also my solution is not safe as the user could change the client-side JS code and ignore the user auth.

Step 1:
Create HTML file for `/dashboard`:

```
app.get('/dashboard', function(req, res) {
      res.sendFile(__dirname + "/" + "index.html");
});
```
and here using this JS code:

```
var headers = new Headers();
var bearer = "Bearer " + "ey...........Ac"; // <---- accessToken
headers.append("Authorization", bearer);
var options = {
        method: "GET",
        headers: headers
};
var graphEndpoint = "https://graph.microsoft.com/v1.0/me";

fetch(graphEndpoint, options)
    .then(resp => {
         // when error redirect to ... otherwise do log:
         console.log(123);
    });
```

on the `<---- accessToken` putting a valid token and it works. Only when token is valid the console.log is done.

So this works, but as I said, if the user is manipulating the code he can still see the page. Also the page is loading until the script is fired. So I cannot see a real value on this. This should happen on server-side somehow. Any idea?

**Comment by pkanher617** [comment]
> It's so confusing for me to have so many different examples and now to pick the best that fits in my case. But I still do not understand at all when to choose the node version and when to choose the browser version.

We will try to update the docs in the main README to be more clear, but generally msal-node is for node.js based apps, while msal-browser is for apps that reside only in the browser.

> But when I see the examples there sometimes the clientSecret is also delievered to the browser to be run on client-side. Isn't this a security issue? Or is the clientSecret not secret at all?

Could you point to the example you are referring to? `clientSecret` should never be sent to the browser AFAIK.

> So as I understand you, node is not needed and I should write my whole code in browser? Or did I get you wrong?

It depends on your app model. Is your application a traditional web application, or a single-page application? If it is a single page application, you should use the msal-browser package. Otherwise, the msal-node package is what you will probably want to use. I recommend reading through the docs for both of these libraries: [msal-browser](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-browser) and [msal-node](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-node#about).

> Could you maybe tell me if that is a proper way?

I'm still confused on what kind of app you are creating. If the user is able to manipulate client-side JS, that sounds like a XSS attack that you will need to ensure can't happen. 

From your original issue, you are saying you are creating a node app. If so, then the only part that requires a browser is getting the code. 

The rest is done in the node.js portion of your app. Once you have the code, you call acquireTokenByCode() to get a token. Then you use that token to query the Graph API. Then you can pass that information to the dashboard page or whatever it is you are trying to display. No tokens should be going to browser-side storage or code.

**Comment by kwoxer** [comment]
Well from Wikipedia I would say my app is no [SPA](https://en.wikipedia.org/wiki/Single-page_application). But it would not be impossible at all.

I read both projects now, but I'm still not sure how to check the user for a valid token in `msal-node`.

I would like to do the following:

> MSAL maintains a token cache (or two caches for confidential client applications) and caches a token after it's been acquired. In many cases, attempting to silently get a token will acquire another token with more scopes based on a token in the cache. It's also capable of refreshing a token when it's getting close to expiration (as the token cache also contains a refresh token).

from the list:

> Confidential Client:
> 
>     Authorization Code Grant with a client credential
>     Refresh Token Grant
>     Silent Flow
>     Client Credential Grant
>     On-behalf-of flow
>     Username and Password flow

So as you said it is possible. But I cannot find a tutorial about this that works.

So how to do the `Silent Flow` in `msal-node`?

**Comment by kwoxer** [comment]
I gave https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/standalone-samples/silent-flow another try. 

But I always get `invalid_client`. What I'm doing wrong?

![image](https://user-images.githubusercontent.com/1167229/109464295-c11ba780-7a66-11eb-9c58-8eed2022a141.png)

**Comment by kwoxer** [comment]
All I need is that the documentation **[here](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/standalone-samples/auth-code)** is extended with the example of a `/dashboard`.

**Comment by kwoxer** [comment]
I search a bit around and found out that I need to set App to Public

![image](https://user-images.githubusercontent.com/1167229/109467426-70f31400-7a6b-11eb-8b91-a9c467f77149.png)

and now it directly worked out

![image](https://user-images.githubusercontent.com/1167229/109467471-836d4d80-7a6b-11eb-847c-2291d2b061b2.png)

**Comment by pkanher617** [comment]
> And why clientSecret needs to be set in msal-node, but not in msal-browser?

Looks like you found the answer above! `msal-browser` is a public client, so it cannot keep secrets. `msal-node` can be either a public or confidential client, so you need to make sure you specify that in the app configuration. 

> So how to do the Silent Flow in msal-node?

I pasted this in my [first response](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/3067#issuecomment-784325823), but you should be able to call acquireTokenSilent on whichever client you create. Let me know if it's still not clear, but the samples linked there should point you in the right direction, as well as the docs for msal-node [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-common/docs/request.md#silent-flow).

> And this should should definitely be better communicated in the docs.

Totally agree, we have added a backlog item to add this to our docs.

**Comment by kwoxer** [comment]
Ok so testing the silent-flow was a good idea. It works great. So there has been a bad setup when I tested it in the past or that Public setting was the reason.

https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/standalone-samples/silent-flow

This works with the `PublicClientApplication` and `acquireTokenSilent` works also as expected.

Anyway the documentation needs to be extended. Because this setup is the best for many web applications but I just found it now after weeks. So making it more prominent would help.
